tests/builtin_server/tests.py

```diff
@@ -89,6 +94,29 @@ def test_file_wrapper_no_sendfile(self):
         self.assertEqual(handler.stdout.getvalue().splitlines()[-1], b'Hello World!')
         self.assertEqual(handler.stderr.getvalue(), b'')
 
+    @override_settings(ROOT_URLCONF='builtin_server.urls')
+    def test_file_response_closing(self):
+        """
+        Returning a `FileResponse` should properly close the file and the http
+        response when `file_wrapper` is used
+        """
+        env = RequestFactory().get('/fileresponse/').environ
+        handler = FileWrapperHandler(None, BytesIO(), BytesIO(), env)
+        handler.run(get_internal_wsgi_application())
+        # Verify that sendfile was used which only happens if file_wrapper got
+        # used.
+        self.assertTrue(handler._used_sendfile)
+        # Fetch the original response object
+        self.assertIn('response', FILE_RESPONSE_HOLDER)
+        response = FILE_RESPONSE_HOLDER['response']
+        # The response should have been closed ...
+        self.assertIs(response.closed, True)
+        # ... as well as all individual file buffers
+        buf1, buf2 = FILE_RESPONSE_HOLDER['buffers']
+        self.assertTrue(buf1.closed)
+        self.assertTrue(buf2.closed)
+        del FILE_RESPONSE_HOLDER['response']
```

Comment:
We should clean `FILE_RESPONSE_HOLDER` in `finally` to ensure tests isolation, e.g.:
```python
try:    
    # Verify that sendfile was used which only happens if file_wrapper got
    # used.
    self.assertTrue(handler._used_sendfile)
    # Fetch the original response object
    self.assertIn('response', FILE_RESPONSE_HOLDER)
    response = FILE_RESPONSE_HOLDER['response']
    # The response should have been closed ...
    self.assertIs(response.closed, True)
    # ... as well as all individual file buffers
    buf1, buf2 = FILE_RESPONSE_HOLDER['buffers']
    self.assertIs(buf1.closed, True)
    self.assertIs(buf2.closed, True)
finally:
    FILE_RESPONSE_HOLDER.pop('response', None)
    FILE_RESPONSE_HOLDER.pop('buffers', None)
```

Reply:
A simple `FILE_RESPONSE_HOLDER.clear()` would do, it is really just for that view ;)

Reply:
🗑️